File: db4on/tutorial/vb/db4o-tutorial-chapters/src/com/db4o/f1/chapter4/InheritanceExample.vb
Imports System
Imports System.IO
Imports com.db4o

Imports com.db4o.query
Namespace com.db4o.f1.chapter4
	Public Class InheritanceExample
	Inherits Util
		Public Shared Sub Main(ByVal args As String())
			File.Delete(Util.YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(Util.YapFileName)
			Try
				StoreFirstCar(db)
				StoreSecondCar(db)
				RetrieveTemperatureReadoutsQBE(db)
				RetrieveAllSensorReadoutsQBE(db)
				RetrieveAllSensorReadoutsQBEAlternative(db)
				RetrieveAllSensorReadoutsQuery(db)
				RetrieveAllObjects(db)
			Finally
				db.Close()
			End Try
		End Sub

		Public Shared Sub StoreFirstCar(ByVal db As ObjectContainer)
			Dim car1 As Car = New Car("Ferrari")
			Dim pilot1 As Pilot = New Pilot("Michael Schumacher", 100)
			car1.Pilot = pilot1
			db.[Set](car1)
		End Sub

		Public Shared Sub StoreSecondCar(ByVal db As ObjectContainer)
			Dim pilot2 As Pilot = New Pilot("Rubens Barrichello", 99)
			Dim car2 As Car = New Car("BMW")
			car2.Pilot = pilot2
			car2.Snapshot()
			car2.Snapshot()
			db.[Set](car2)
		End Sub

		Public Shared Sub RetrieveAllSensorReadoutsQBE(ByVal db As ObjectContainer)
			Dim proto As SensorReadout = New SensorReadout(DateTime.MinValue, Nothing, Nothing)
			Dim result As ObjectSet = db.[Get](proto)
			ListResult(result)
		End Sub

		Public Shared Sub RetrieveTemperatureReadoutsQBE(ByVal db As ObjectContainer)
			Dim proto As SensorReadout = New TemperatureSensorReadout(DateTime.MinValue, Nothing, Nothing, 0)
			Dim result As ObjectSet = db.[Get](proto)
			ListResult(result)
		End Sub

		Public Shared Sub RetrieveAllSensorReadoutsQBEAlternative(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.[Get](GetType(SensorReadout))
			ListResult(result)
		End Sub

		Public Shared Sub RetrieveAllSensorReadoutsQuery(ByVal db As ObjectContainer)
			Dim query As Query = db.Query()
			query.Constrain(GetType(SensorReadout))
			Dim result As ObjectSet = query.Execute()
			ListResult(result)
		End Sub

		Public Shared Sub RetrieveAllObjects(ByVal db As ObjectContainer)
			Dim result As ObjectSet = db.[Get](New Object())
			ListResult(result)
		End Sub

	End Class
End Namespace

File: db4on/tutorial/db4o-tutorial-chapters/src/com/db4o/f1/chapter4/InheritanceExample.cs
using System;
using System.IO;
using com.db4o;
using com.db4o.f1;
using com.db4o.query;

namespace com.db4o.f1.chapter4
{   
    public class InheritanceExample : Util
    {        
        public static void Main(string[] args)
        {
            File.Delete(Util.YapFileName);          
            ObjectContainer db = Db4o.OpenFile(Util.YapFileName);
            try
            {
                StoreFirstCar(db);
                StoreSecondCar(db);
                RetrieveTemperatureReadoutsQBE(db);
                RetrieveAllSensorReadoutsQBE(db);
                RetrieveAllSensorReadoutsQBEAlternative(db);
                RetrieveAllSensorReadoutsQuery(db);
                RetrieveAllObjects(db);
            }
            finally
            {
                db.Close();
            }
        }
        
        public static void StoreFirstCar(ObjectContainer db)
        {
            Car car1 = new Car("Ferrari");
            Pilot pilot1 = new Pilot("Michael Schumacher", 100);
            car1.Pilot = pilot1;
            db.Set(car1);
        }
        
        public static void StoreSecondCar(ObjectContainer db)
        {
            Pilot pilot2 = new Pilot("Rubens Barrichello", 99);
            Car car2 = new Car("BMW");
            car2.Pilot = pilot2;
            car2.Snapshot();
            car2.Snapshot();
            db.Set(car2);
        }
        
        public static void RetrieveAllSensorReadoutsQBE(ObjectContainer db)
        {
            SensorReadout proto = new SensorReadout(DateTime.MinValue, null, null);
            ObjectSet result = db.Get(proto);
            ListResult(result);
        }
        
        public static void RetrieveTemperatureReadoutsQBE(ObjectContainer db)
        {
            SensorReadout proto = new TemperatureSensorReadout(DateTime.MinValue, null, null, 0.0);
            ObjectSet result = db.Get(proto);
            ListResult(result);
        }
        
        public static void RetrieveAllSensorReadoutsQBEAlternative(ObjectContainer db)
        {
            ObjectSet result = db.Get(typeof(SensorReadout));
            ListResult(result);
        }
        
        public static void RetrieveAllSensorReadoutsQuery(ObjectContainer db)
        {
            Query query = db.Query();
            query.Constrain(typeof(SensorReadout));
            ObjectSet result = query.Execute();
            ListResult(result);
        }
        
        public static void RetrieveAllObjects(ObjectContainer db)
        {
            ObjectSet result = db.Get(new object());
            ListResult(result);
        }
    }
}
